### PR TITLE
LPD-31008 Cannot change the source of a selection filter on edit

### PR DESCRIPTION
**Pull request by markocikos**
### References

- Parent Story: [LPD-12345 Fix all the things](https://issues.liferay.com/browse/LPD-12345)
- This is a followup on https://github.com/liferay-frontend/liferay-portal/pull/4361

### What is the goal of this PR?

In the PR:
- We are restructuring selection filters handling, squashing unnecessary `useEffects`. On the way, we are fixing behaviors on source selection. This is a complete rework of the original solution.
- We are fixing few flakiness issues, on login and on fragment selection

Few technical notes inline.

### What does it look like?

https://github.com/user-attachments/assets/ebd0c530-74dd-4ef1-9600-3170fa18161f

image

### CODE REVIEW

File: modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java [hunk 1]
```diff
@@ -939,11 +937,12 @@ private JSONArray _getSelectedItemsJSONArray(
...
 
 		for (int i = 0; i < preselectedValuesJSONArray.length(); i++) {
-			String key = preselectedValuesJSONArray.getString(i);
+			JSONObject jsonObject = preselectedValuesJSONArray.getJSONObject(i);
 
 			for (ListTypeEntry listTypeEntry : listTypeEntries) {
 				if (Objects.equals(
-						listTypeEntry.getExternalReferenceCode(), key)) {
+						listTypeEntry.getExternalReferenceCode(),
+						jsonObject.getString("value"))) {
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/fragment/renderer/FDSAdminFragmentRenderer.java:945]: `preselectedValues` db field / state variable had two formats, for two source types. One was an array of label/value objects, the other array of strings (ERCs). This complicated code. Now it's always objects. We are sacrificing a tiny bit of performance, as we have a slightly bigger footprint in DB that necessary, but I think it's worth it for the code simplification.

File: modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/resources/META-INF/resources/js/data_set/filters/modals/selection_filter/SelectionFilter.tsx [hunk 2]
```diff
@@ -221,182 +216,39 @@ function Body({
...
-							.filter((item) =>
-								fuzzy.match(preselectedValueInput, item.name)
-							)
-							.map((item) => ({
-								label: item.name,
-								value: String(item.externalReferenceCode),
-							}))
-			);
-		}
-	}, [preselectedValueInput, source, sourceType]);
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/resources/META-INF/resources/js/data_set/filters/modals/selection_filter/SelectionFilter.tsx:385]: :skull:

File: modules/test/playwright/utils/performLogin.ts [hunk 3]
```diff
@@ -46,12 +46,18 @@ async function performLogin(
 
 	await page.goto(baseUrl);
 
-	const signInButton = await page.getByRole('button', {name: 'Sign In'});
+	const signInButton = page.getByRole('button', {name: 'Sign In'});
+
+	await expect(page.getByPlaceholder('Search')).toBeVisible();
```
Comment (markocikos) [inline modules/test/playwright/utils/performLogin.ts:51]: This fix flakiness in login. Flakiness is caused by the fact that `Sign In` button is not usable for a short time when the page is first loaded. We are now waiting for `Search` input to be visible, using it as an indicator that hydration completed.

**Comment by markocikos** [comment]
> The commit message contains the word 'login', which could reveal potential security vulnerablities.

:facepalm:

**Comment by markocikos** [comment]
ci:test:sf

**Comment by markocikos** [comment]
ci:test:relevant

File: modules/test/playwright/utils/performLogin.ts (hunk 3 above)
Comment (izaera) [reply modules/test/playwright/utils/performLogin.ts:51]: :+1: The previous code was copied from the very first PoC so I assumed it was correct. If it was flaky and this change makes waiting more reliable it's great :+1:

File: modules/test/playwright/utils/performLogin.ts [hunk 4]
```diff
@@ -60,7 +66,7 @@ async function performLogin(
 	}
 	else {
 		await page
-			.getByLabel('Sign In- Loading')
+			.locator('.login-container')
```
Comment (izaera) [inline modules/test/playwright/utils/performLogin.ts:69]: Mmm, not sure about this change. I remember having seen some modifications from other developer to cover a different use case. Maybe this breaks that fix :thinking: 

I'm referring to the latest three commits for this file -> https://github.com/liferay/liferay-portal/commits/master/modules/test/playwright/utils/performLogin.ts

Comment (markocikos) [reply modules/test/playwright/utils/performLogin.ts:69]: I can revert this bit, I thought it minor in importance. I simply don't like when `getByLabel` includes parts that are not visible in UI. Some ugliness in needed in either case as there are two `Sign In` buttons in UI (main page and within modal). Maybe there is something here that I'm not seeing, maybe there is a case when there is no `login-container`, so I'll revert it.

Comment (markocikos) [reply modules/test/playwright/utils/performLogin.ts]: Reverted!

**Comment by markocikos** [comment]
ci:forward

**Pull request closed without merging**